Feature/BJS2-41788 Notification Service: Recommendation Received

drewbal opened the pull request.

Commit:
Feature: Recommendation Received Notification: Notification Service setup

Redis configuration added;

Commit:
Feature: Recommendation Received Notification: Notification Service setup

updates Redis config;
adds MessageSource config;
updates MessageBuilder interface;
updates messages.properties;

Commit:
Feature: Recommendation Received Notification: Listener setup

adds UserContactsDto;
updates userServiceClient: connection to new endpoint;
updates NotificationService interface;

Commit:
Feature: Recommendation Received Notification: Listener setup

adds RecommendationReceivedEvent dto;
adds Listener;
adds MessageBuilder;

Commit:
Merge branch 'Dragon-master-stream7' into feature/BJS2-41788-Recommendation-Notification

# Conflicts:
#	src/main/java/faang/school/notificationservice/NotificationServiceApp.java
#	src/main/java/faang/school/notificationservice/dto/UserContactsDto.java
#	src/main/java/faang/school/notificationservice/service/NotificationService.java

Commit:
Feature: Recommendation Received Notification: Conflicts solved

DeemaP commented on src/main/java/faang/school/notificationservice/dto/UserContactsDto.java:
```diff
@@ -14,8 +14,7 @@
 @NoArgsConstructor
-public class UserDto {
-
+public class UserContactsDto {
```
Не знаю почему не могу оставить комменты ниже по коду в этой дто, напишу сверху тогда.
@PositiveOrZero(message = "Id must be positive or zero")
кажется не очень аннотация для id, они вроде у нас все просто @Positive


Наверное не сильно важно, но не стоит ли regexp из StringPhone вынести в константу, а не хардкодить?
(тут конечно можно мне ответить, что все сообщения в аннотациях хардкод) Просто интересуюсь твоим мнением.

drewbal replied:
Хорошие вопросы. Я их пропустил, если честно, потому что это проверенный код с мастера :) 
@PositiveOrZero я уже исправил, а вот регулярка дает подумать. 
В целом я с тобой согласен, но есть нюанс. Аннотация требует именно стрингу в нее. Вынесение в константу или конфиг не работает

DeemaP replied:
Ну тогда хардкод))

DeemaP commented on src/test/java/faang/school/notificationservice/messaging/RecommendationMessageBuilderTest.java:
```diff
@@ -0,0 +1,61 @@
+
+        Class<?> result = messageBuilder.getInstance();
+
+        // Assert that the method returns null
```
коммент потри)

Commit:
Feature: Recommendation Received Notification:

Cross-review fixes

RuslanZinovyev commented on src/main/java/faang/school/notificationservice/listener/RecommendationReceivedEventListener.java:
```diff
@@ -0,0 +1,51 @@
+            RecommendationReceivedEvent event = objectMapper.readValue(message.getBody(), RecommendationReceivedEvent.class);
+            handleEvent(event);
+        } catch (IOException e) {
+            log.error("Error while serializing {} from redis. Error: {}", message.getBody(), e.getMessage(), e);
```
Я желаю удачи кому-нибудь прочитать содержимое message.getBody() в логе, что по сути будет массивом из байт
Если ты хочешь видеть содержимое то надо конвертнуть в читаемый формат, например String

drewbal replied:
Мы ж программисты! Байты наше всё! Слава Роботам! Сжечь всех кожаные мешки!

Но ладно, дадим этим слабым человекам последний шанс и переведем в удобный для их примитивного мозга формат :)

RuslanZinovyev commented on src/main/java/faang/school/notificationservice/listener/RecommendationReceivedEventListener.java:
```diff
@@ -0,0 +1,51 @@
+        UserContactsDto receiverDto = userServiceClient.getUserContacts(event.getReceiverId());
+        UserContactsDto authorDto = userServiceClient.getUserContacts(event.getAuthorId());
+
+        Object[] names = {receiverDto.getUsername(), authorDto.getUsername()};
```
Мы на 100% уверены что `receiverDto `и `authorDto `не будут null? А может внешний вызов например выбросить исключение если например сервис не доступен? Что будет с программой в таком случае? Я просто спрашиваю.

drewbal replied:
Очень хороший вопрос. Согласно дизайну взаимодействия наших сервисов, эти юзеры должны быть в базе, если программа дошла до этой точки. Их валидация произошла на более ранних этапах. Но есть риск взаимодействия по http. Для этого я сделал проверку с исключением, чтобы получить необходимые логи.

RuslanZinovyev commented on src/main/java/faang/school/notificationservice/listener/RecommendationReceivedEventListener.java:
```diff
@@ -0,0 +1,51 @@
+
+    private void handleEvent(RecommendationReceivedEvent event) {
+        UserContactsDto receiverDto = userServiceClient.getUserContacts(event.getReceiverId());
+        UserContactsDto authorDto = userServiceClient.getUserContacts(event.getAuthorId());
```
Мне не нравится идея делать внешние вызовы внутри обработчика события, плохо влияет на производительность и в целом не самое удачное решение. Да и вызовы блокируют выполнение.
Можем ли мы получить `RecommendationReceivedEvent` с уже необоходимыми полями
`receiverUsername ` и `authorUsername`?

drewbal replied:
Отличная идея, лежащая на поверхности, но я полез усложнять :) конечно, мы это можем вытащить на этапе формирования события и отправить данные. Тем более, что тут они нужны исключительно для формирования сообщения.

RuslanZinovyev commented on src/main/java/faang/school/notificationservice/messaging/RecommendationMessageBuilder.java:
```diff
@@ -0,0 +1,24 @@
+
+    @Override
+    public Class<?> getInstance() {
+        return null;
```
Возвращать null плохая практика
`
return RecommendationReceivedEvent.class;`

drewbal replied:
Absolutely

RuslanZinovyev commented on src/main/java/faang/school/notificationservice/messaging/RecommendationMessageBuilder.java:
```diff
@@ -0,0 +1,24 @@
+    }
+
+    @Override
+    public String buildMessage(RecommendationReceivedEvent event, Locale locale, Object[] placeholders) {
```
Можем мы сделать что-нибудь наподобие примера ниже, это как пример ты можешь передать в массив все что тебе необходимо

```
@Override
    public String buildMessage(RecommendationReceivedEvent event, Locale locale) {
        return messageSource.getMessage(
            "recommendation.new", 
            new Object[]{event.getReceiverUsername(), event.getAuthorUsername()}, 
            locale
        );
    }
```

drewbal replied:
Согласен. Так даже проще. Опять усложнил

RuslanZinovyev commented on src/main/java/faang/school/notificationservice/messaging/MessageBuilder.java:
```diff
@@ -6,5 +6,5 @@ public interface MessageBuilder<T> {
     Class<?> getInstance();
 
-    String buildMessage(T event, Locale locale);
+    String buildMessage(T event, Locale locale, Object[] placeholders);
```
Меняя интерфейс напрямую ты нарушаешь контракт, этот интерфейс уже может быть использован в коде. Плюс в этом нет необходиомсти, попробуй сделать так чтобы placeholders тут не был нужен изначально.

drewbal replied:
Да, мне приходилось код править в других местах, меняя контракт. Все оказалось проще, вернул и сделал без лишних ненужности

RuslanZinovyev commented on src/main/java/faang/school/notificationservice/service/EmailService.java:
```diff
@@ -15,7 +15,7 @@ public class EmailService implements NotificationService  {
 
     @Override
-    public void send(UserDto user, String message) {
+    public void send(UserContactsDto user, String message) {
```
Я знаю что это не твои изменения, я не заметил когда первый раз смотрел метод send() у emailSender объекта выбрасывает `MailException`, не `EmailSendingException `поэтому в catch() нужно поместить именно `MailException`

drewbal replied:
Да, логично. Я поместил `MailException` в catch с дальнейшим прокидыванием `EmailSendingException`, как было изначально

Commit:
Feature: Recommendation Received Notification:

Review fixes:
Adds Retry for FeignClient
Optimizes notification builder method
Fixed Telegram class (wrong validation annotations)

Commit:
Merge branch 'Dragon-master-stream7' into feature/BJS2-41788-Recommendation-Notification

# Conflicts:
#	src/main/resources/application.yaml

RuslanZinovyev commented on src/main/java/faang/school/notificationservice/listener/RecommendationReceivedEventListener.java:
```diff
@@ -0,0 +1,78 @@
+            return userServiceClient.getUserContacts(userId);
+        } catch (FeignException e) {
+            log.error("Error occurred while fetching user contacts for user {}", userId, e);
+            return null;
```
Не рекомендую возвращать null, выброси кастомное исключение и позаботься о нем.

drewbal replied:
Согласен с тобой, можно же проще и лучше. Я решил вот так выбросить кастомное исключение, обработав его в вызывающем методе. Обработка включает в себя только примитивное логирование и причину.

Commit:
Feature: Recommendation Notification : Notification Service

Commit:
Feature: Recommendation Notification : Notification Service

Review fix
